### PR TITLE
Add VersionId parameter for S3 reads

### PR DESCRIPTION
**Pull request by kukushking**
*Issue #721 *

*Description of changes:*
Add `version_id: Optional[Union[str, Dict[str, str]]]` parameter to S3 read operations to be able to query specific versions of an object or mulptiple objects. The parameter, depending on the context of the operation, may be either a single version id string or a dictionary `object key -> version id`.

Tests coming up.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by jaidisido** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: bafac25067c5b2b57f7ec0230dcd97f2881079cb
* Result: FAILED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=27384e1b-074c-44c5-89e3-46fa107c854d%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jaidisido** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: 51411018e8e3c1b6595f0586924ed5444657089b
* Result: SUCCEEDED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=e4a5e7e7-3db0-4a2e-a346-b5171d43b2a2%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jaidisido** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: 5170150723dd1754ffd9a72f44494a96f87fc213
* Result: SUCCEEDED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=31d791f1-ffe0-4306-9c72-b9710990dfc1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jaidisido** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: d817504574afb859def541436bdb8764f7672ff7
* Result: SUCCEEDED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=71b2bbed-2825-4dfc-80fa-b0761394bb36%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

File: awswrangler/s3/_fs.py [hunk 1]
```diff
@@ -106,19 +110,34 @@ def _fetch_range(
...
+        resp = _utils.try_it(
+            f=s3_client.get_object,
+            ex=_S3_RETRYABLE_ERRORS,
+            base=0.5,
+            max_num_tries=6,
+            Bucket=bucket,
+            Key=key,
+            Range=f"bytes={start}-{end - 1}",
+            **boto3_kwargs,
+        )
```
Comment (jaidisido) [inline awswrangler/s3/_fs.py:140]: Same as first comment

File: awswrangler/s3/_read_parquet.py [hunk 2]
```diff
@@ -666,20 +696,30 @@ def read_parquet(
...
+            chunked=chunked,
+            validate_schema=validate_schema,
+            ignore_index=ignore_index,
+            version_ids=versions,
+            **args,
         )
     if len(paths) == 1:
-        return _read_parquet(path=paths[0], **args)
+        return _read_parquet(
+            path=paths[0], **args, version_id=versions[paths[0]] if isinstance(versions, dict) else None
```
Comment (jaidisido) [inline awswrangler/s3/_read_parquet.py:708]: Being picky here but can we put **args as the last argument?

File: tests/test_s3_excel.py [hunk 3]
```diff
@@ -17,3 +18,19 @@ def test_excel(path, ext, use_threads):
...
+    bucket, key = wr._utils.parse_path(path_file)
+    wr.s3.to_excel(df=df1, path=path_file, index=False)
+    wr.s3.to_excel(df=df2, path=path_file, index=False)
+    object_versions = [
+        v["VersionId"] for v in boto3.client("s3").list_object_versions(Bucket=bucket, Prefix=key)["Versions"]
+    ]
+    assert len(object_versions) == 2
+    for object_version, df in zip(object_versions, [df2, df1]):
+        df_temp = wr.s3.read_excel(path_file, version_id=object_version)
+        assert df_temp.equals(df)
```
Comment (jaidisido) [inline tests/test_s3_excel.py:36]: How about:
```
def test_read_xlsx_versioned(path) -> None:
    path_file = f"{path}0.xlsx"
    dfs = [pd.DataFrame({"c0": [0, 1, 2], "c1": [3, 4, 5]}), pd.DataFrame({"c0": [3, 4, 5], "c1": [6, 7, 8]})]

    for df in dfs:
        wr.s3.to_excel(df=df, path=path_file, index=False)
        version_id = wr.s3.describe_objects(path=path_file)[path_file]["VersionId"]
        df_temp = wr.s3.read_excel(path_file, version_id=version_id)
        assert df_temp.equals(df)
        assert version_id == wr.s3.describe_objects(path=path_file, version_id=version_id)[path_file]["VersionId"]
```
This way we also test `wr.s3.describe_objects` for more coverage and avoid using boto3 altogether

File: awswrangler/s3/_list.py [hunk 4]
```diff
@@ -187,7 +190,10 @@ def does_object_exist(
     else:
         extra_kwargs = {}
     try:
-        client_s3.head_object(Bucket=bucket, Key=key, **extra_kwargs)
+        if version_id:
+            client_s3.head_object(Bucket=bucket, Key=key, VersionId=version_id, **extra_kwargs)
+        else:
+            client_s3.head_object(Bucket=bucket, Key=key, **extra_kwargs)
```
Comment (jaidisido) [inline awswrangler/s3/_list.py:196]: Same as first comment

**Review by jaidisido — COMMENTED** [review]
Looking great, thank you for this. Added a few suggestions

File: awswrangler/s3/_describe.py [hunk 5]
```diff
@@ -28,21 +31,38 @@ def _describe_object(
...
+            ex=client_s3.exceptions.NoSuchKey,
+            Bucket=bucket,
+            Key=key,
+            VersionId=version_id,
+            **extra_kwargs,
+        )
+    else:
+        desc = _utils.try_it(
+            f=client_s3.head_object, ex=client_s3.exceptions.NoSuchKey, Bucket=bucket, Key=key, **extra_kwargs
+        )
```
[the remark this answers is not in the document]
Comment (kukushking) [reply awswrangler/s3/_describe.py]: Agree, that's much cleaner, thanks.

File: awswrangler/s3/_download.py [hunk 6]
```diff
@@ -14,6 +14,7 @@
 def download(
     path: str,
     local_file: Union[str, Any],
+    version_id: Optional[str] = None,
```
[the remark this answers is not in the document]
Comment (kukushking) [reply awswrangler/s3/_download.py:17]: Oops

**Comment by jaidisido** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: 5da57462de600f54e17db03fce434e752ad265c4
* Result: FAILED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bf5be575-cbaa-4d83-9383-e6951e0a3527%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jaidisido** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: beb78b150ac917734ea3c0498f480442532d215f
* Result: FAILED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1836ab0e-0295-40bd-b014-fba0b69c637c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Comment by jaidisido** [comment]
### AWS CodeBuild CI Report

* CodeBuild project: GitHubCodeBuild8756EF16-pDO66x4b9gEu
* Commit ID: 6f3142628b2a703ccd530ec544ec3161a8ba794a
* Result: SUCCEEDED
* [Build Logs](https://8z22lm66x0.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=df5c0213-a636-4a79-9660-bdcc5b287667%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

**Pull request merged**